src/main/kotlin/com/exactpro/th2/codec/json/JsonPipelineCodec.kt

```diff
@@ -10,104 +26,106 @@ import com.exactpro.th2.common.grpc.MessageGroup
         val messages = messageGroup.messagesList
+        val builder = MessageGroup.newBuilder()
 
-        if (messages.isEmpty()) {
-            return messageGroup
-        }
-
-        require(messages.size == 1) { "Message group must have one value." }
-
-        val message = messages[0].message
+        for (message in messages) {
+            if (!message.hasMessage()) {
+                builder.addMessages(message)
+                continue
+            }
+            if (message.message.metadata.run { protocol.isNotEmpty() && protocol != PROTOCOL }) {
+                builder.addMessages(message)
+                continue
+            }
 
-        require(message.metadata.protocol == PROTOCOL) { "First message in group must have $PROTOCOL type." }
+            val parsedMessage = message.message
+            val metadata = parsedMessage.metadata
 
-        val builder = MessageGroup.newBuilder()
-        val metadata = message.metadata
+            builder += when(val direction = parsedMessage.direction) {
+                FIRST, SECOND -> writer.writeValueAsString(parsedMessage.toMap())
+                else -> error("Unsupported message direction: $direction")
+            }.toByteArray(UTF_8).toRawMessage(
+                metadata.id,
+                metadata.propertiesMap,
+                parsedMessage.parentEventId
+            )
 
-        builder += when(val direction = message.direction) {
-            FIRST, SECOND -> {
-                MAPPER.writeValueAsString(message.toMap())
-            }
-            else -> error("Unsupported message direction: $direction")
-        }.toByteArray(UTF_8).toRawMessage(
-            metadata.id,
-            metadata.propertiesMap,
-            message.parentEventId
-        )
+        }
 
         return builder.build()
     }
 
     companion object {
-        const val REQUEST = "Request"
-        const val RESPONSE = "Response"
-        val MAPPER = ObjectMapper()
-
-        fun injectCustomSerializer() {
+        private const val INCOMING = "Incoming"
+        private const val OUTGOING = "Outgoing"
+        private val MAPPER = ObjectMapper().apply {
             val module = SimpleModule()
             module.addSerializer(String::class.java, TypeInfoSerializer())
-            MAPPER.registerModule(module)
+            module.addDeserializer(Any::class.java, TypeInfoDeserializer())
+            this.registerModule(module)
         }
 
-        fun injectCustomDeserializer() {
-            val module = SimpleModule()
-            module.addDeserializer(Any::class.java, TypeInfoDeserializer())
-            MAPPER.registerModule(module)
+        fun getReader(appendTypeInfo: Boolean): ObjectReader {
+            return MAPPER.reader().withAttribute(DECODE_TYPE_INFO_PROPERTY, appendTypeInfo)
+        }
+
+        fun getWriter(unpackTypeInfo: Boolean): ObjectWriter {
+            return MAPPER.writer().withAttribute(ENCODE_TYPE_INFO_PROPERTY, unpackTypeInfo)
         }
```

Comment:
I think it would be better to just create a separate mapper for each codec instance. This way we'll avoid adding unneccessary serializer/deserlializer in case of disabled option. There won't be much profit in them anyway - they'll be just checking for presence of our attribute on every read/write and that's all